bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/discovery/SomfyMyLinkDeviceDiscoveryService.java

```diff
@@ -0,0 +1,196 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.somfymylink.internal.SomfyMyLinkHandlerFactory;
+import org.openhab.binding.somfymylink.internal.handler.SomfyMyLinkBridgeHandler;
+import org.openhab.binding.somfymylink.internal.handler.SomfyMyLinkException;
+import org.openhab.binding.somfymylink.internal.model.SomfyMyLinkScene;
+import org.openhab.binding.somfymylink.internal.model.SomfyMyLinkShade;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SomfyMyLinkDeviceDiscoveryService} is responsible discovering things connected to the mylink.
+ *
+ * @author Chris Johnson - Initial contribution
+ */
+@NonNullByDefault
+public class SomfyMyLinkDeviceDiscoveryService extends AbstractDiscoveryService {
+
+    private static final int DISCOVERY_REFRESH_SEC = 1800;
+
+    private final Logger logger = LoggerFactory.getLogger(SomfyMyLinkDeviceDiscoveryService.class);
+
+    private @Nullable ScheduledFuture<?> scanTask;
+
+    private SomfyMyLinkBridgeHandler mylinkHandler;
+
+    private @Nullable ScheduledFuture<?> discoveryJob;
+
+    public SomfyMyLinkDeviceDiscoveryService(SomfyMyLinkBridgeHandler mylinkHandler) throws IllegalArgumentException {
+        super(SomfyMyLinkHandlerFactory.DISCOVERABLE_DEVICE_TYPES_UIDS, 10);
+
+        this.mylinkHandler = mylinkHandler;
+    }
+
+    @Override
+    @Activate
+    public void activate(@Nullable Map<@NonNull String, @Nullable Object> configProperties) {
+        super.activate(configProperties);
+    }
+
+    @Override
+    @Deactivate
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting Somfy My Link background discovery");
+
+        if (discoveryJob == null || discoveryJob.isCancelled()) {
+            discoveryJob = scheduler.scheduleWithFixedDelay(this::runDiscovery, 10, DISCOVERY_REFRESH_SEC,
+                    TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stopping Somfy MyLink background discovery");
+        if (discoveryJob != null && !discoveryJob.isCancelled()) {
+            discoveryJob.cancel(true);
+            discoveryJob = null;
+        }
```

Comment:
There is no harm in trying to cancel an already canceled task. Also, in order to deal with the null check warnings and to prevent any concurrency mishaps you should cache fields in a local variable.
```suggestion
        ScheduledFuture<?> discoveryJob = this.discoveryJob;
        if (discoveryJob != null ) {
            discoveryJob.cancel(true);
            this.discoveryJob = null;
        }
```

Reply:
done